Lib/asyncio/unix_events.py

```diff
@@ -1101,6 +1108,118 @@ def _do_waitpid_all(self):
                 callback(pid, returncode, *args)
 
 
+class MultiLoopChildWatcher(AbstractChildWatcher):
+    # The class keeps compatibility with AbstractChildWatcher ABC
+    # To achieve this it has empty attach_loop() method
+    # and don't accept explicit loop argument
+    # for add_child_handler()/remove_child_handler()
+    # but retrieves the current loop by get_running_loop()
+
+    def __init__(self):
+        self._callbacks = {}
+        self._saved_sighandler = None
+
+    def close(self):
+        self._callbacks.clear()
+        if self._saved_sighandler is not None:
+            signal.signal(signal.SIGCHLD, self._saved_sighandler)
+            self._saved_sighandler = None
+        super().close()
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, a, b, c):
+        # SafeChildWatcher doesn't close on __exit__ too
+        pass
+
+    def add_child_handler(self, pid, callback, *args):
+        loop = tasks.get_running_loop()
+        self._callbacks[pid] = (weakref.ref(loop), callback, args)
+
+        # Prevent a race condition in case the child is already terminated.
+        self._do_waitpid(pid)
+
+    def remove_child_handler(self, pid):
+        try:
+            del self._callbacks[pid]
+            return True
+        except KeyError:
+            return False
+
+    def attach_loop(self, loop):
+        # Don't attach the loop but initialize itself if called first time
+        # The reason to do it here is that attach_loop() is called from
+        # unix policy only for the main thread.
+        # Main thread is required for subscription on SIGCHLD signal
```

Comment:
Can you add an assertion that the current thread is indeed the main thread?

Reply:
The assertion is not needed: `signal.signal` raises an exception if the thread is not main.

Reply:
Right, never mind.